### PR TITLE
Eraser

### LINKED ISSUE #30: feature idea: rubber to drawing area
when the rubber touches a drawn object, the entire object gets deleted, not just a part of it (otherwise rubbering a circle too difficult to implement)

rubbering must be undoable

### PR DESCRIPTION
**Pull request by Akuli**
Closes #30 

TODO:
- [x] add rubber icon
- [x] when hovering canvas with rubber selected, show the circle around mouse cursor that will be rubbered
- [x] rubbering part of a line without rubbering the entire line
- [x] different sizes of rubbers

### CODE REVIEW

**Comment by Akuli** [comment]
I might work on this PR this weekend

**Comment by Akuli** [comment]
...or maybe next week lol

**Comment by Akuli** [comment]
Should it be called rubber or eraser?

**Comment by PurpleMyst** [comment]
I'm a bigger fan of "eraser" because "rubber" has different connotations to me

**Comment by Akuli** [comment]
I'd like a review, even though it's still missing some features

**Comment by Akuli** [comment]
note to self: undo is still buggy

**Comment by Akuli** [comment]
@PurpleMyst Can you review?

File: js/drawobjects/pen.ts [hunk 1]
```diff
@@ -9,10 +9,62 @@ that was still true after compressing with lzstring, with almost nothing drawn
...
+  const squareRoot = Math.sqrt(ACBA*ACBA - ACAC*BABA + BABA*radius*radius);
+  if (isNaN(squareRoot)) {
+    // No intersection points of circle and line
+    return null;
+  }
+  const tMin = (-ACBA - squareRoot)/BABA;
+  const tMax = (-ACBA + squareRoot)/BABA;
+
+  const result = intervalSetDifference([0, 1], [tMin, tMax]).map(tInterval => tInterval.map(t => f(t)) as [Point, Point]);
+  if (result + "" == [[A, B]] + "") {
```
Comment (PurpleMyst) [inline js/drawobjects/pen.ts:62]: Is there a better way of doing this than exploting JavaScript's weak typing?

File: js/drawobjects/pen.ts [hunk 2]
```diff
@@ -24,25 +76,51 @@ export class Pen implements DrawObject {
...
+      return [];
+    }
+    if (!somethingChanged) {
+      return [this];
+    }
+
+    const resultLines = [new Pen(lineSegments[0][0], this.color)];
+    for (const [start, end] of lineSegments) {
+      let lastLine = resultLines[resultLines.length - 1];
+      if (start + "" !== lastLine.points[lastLine.points.length - 1] + "") {
```
Comment (PurpleMyst) [inline js/drawobjects/pen.ts:117]: See above. I don't like this

**Review by PurpleMyst — CHANGES_REQUESTED** [review]
Here's my review, my concerns are in individual comments :)
Apart from the few nitpicks, very nice code! Great job

File: js/drawobjects/pen.ts (hunk 1 above)
Comment (Akuli) [reply js/drawobjects/pen.ts:62]: This code is comparing nested number arrays by value. It's intended to do this, but in a much less ugly way:

```js
if (result.length === 1 && result[0][0][0] === A[0] && result[0][0][1] === A[1] && result[0][1][0] === B[0] && result[0][1][1] === B[1]) {
```

What do you think about defining a `pointsEqual` function and then doing something like this?

```js
if (result.length === 1 && pointsEqual(result[0][0], A) && pointsEqual(result[0][1], B)) {
```

File: js/canvasmanager.ts [hunk 3]
```diff
@@ -42,24 +89,28 @@ export class CanvasManager extends StrictEventEmitter<CanvasManagerEvents>() {
 
   objects: DrawObject[] = [];
 
-  private events: Event[] = [];
+  // each splice list should be in the order that changes happened (will be reversed when undoing)
+  private undoSpliceLists: Splice<DrawObject>[][] = [];
```
[the remark this answers is not in the document]
Comment (Akuli) [reply js/canvasmanager.ts:93]: How about `undoStack`?

[the remark this answers is not in the document]
Comment (Akuli) [reply js/canvasmanager.ts]: Yes. I documented it better.

File: js/canvasmanager.ts [hunk 4]
```diff
@@ -12,27 +12,74 @@ interface CanvasManagerEvents {
   change: () => void;
 }
 
-interface Event {
-  undo: (cm: CanvasManager) => void;
+type Splice<T> = {
```
[the remark this answers is not in the document]
Comment (Akuli) [reply js/canvasmanager.ts]: Not generic anymore

File: js/canvasmanager.ts [hunk 5]
```diff
@@ -81,37 +132,45 @@ export class CanvasManager extends StrictEventEmitter<CanvasManagerEvents>() {
...
-      this.events.push(new DrawEvent());
+      this.undoSpliceLists.push(this.tool.onMouseDown(this, point));
     });
 
     this.canvas.addEventListener("mousemove", event => {
+      const point = this.xyFromEvent(event);
+      if (point === null) return;
+
+      if (this.tool instanceof Eraser) {
+        (this.tool as Eraser).updateIndicator(this.eraserIndicator, point);
```
[the remark this answers is not in the document]
Comment (Akuli) [reply js/canvasmanager.ts]: Deleted

File: js/drawobjects/drawobject.ts [hunk 6]
```diff
@@ -12,4 +16,7 @@ export interface DrawObject {
 
   onMouseMove(point: Point): void;
   toStringPart(): string;
+
+  // return value is list of objects to replace this with
+  getErasingObjects(eraserCenter: Point, eraserRadius: number): DrawObject[];
```
[the remark this answers is not in the document]
Comment (Akuli) [reply js/drawobjects/drawobject.ts:21]: How about `getReplacingObjectsForErasing`?

File: js/canvasmanager.ts [hunk 7]
```diff
@@ -81,37 +132,45 @@ export class CanvasManager extends StrictEventEmitter<CanvasManagerEvents>() {
...
+      if (this.undoSpliceLists.length !== 0 && this.undoSpliceLists[this.undoSpliceLists.length - 1].length === 0) {
+        this.undoSpliceLists.pop();
+      }
 
-      if (!mouseMoved) {
+      if (!mouseMoved && (this.tool instanceof DrawingTool)) {
         // Draw a dot instead of empty stuff.
         this.objects.pop();
-        const dot = new Circle(pointOrNull!, this.color, true, 2);
+        const dot = new Circle(point!, this.color, true, 2);
```
[the remark this answers is not in the document]
Comment (Akuli) [reply js/canvasmanager.ts]: Deleted

File: js/canvasmanager.ts [hunk 8]
```diff
@@ -12,27 +12,76 @@ interface CanvasManagerEvents {
   change: () => void;
 }
 
-interface Event {
-  undo: (cm: CanvasManager) => void;
+// Represents arguments of array splice method.
```
Comment (PurpleMyst) [inline js/canvasmanager.ts:15]: Could we make this a JSDoc comment `/** ... */` so that it shows up in IDE autocomplete suggestions? (and all other comments introduced in this commit that are meant to document what something is/does)

File: js/canvasmanager.ts [hunk 9]
```diff
@@ -138,45 +144,46 @@ export class CanvasManager extends StrictEventEmitter<CanvasManagerEvents>() {
...
         // Draw a dot instead of empty stuff.
         this.objects.pop();
-        const dot = new Circle(point!, this.color, true, 2);
+        const dot = new Circle(point, this.color, true, 2);
         this.objects.push(dot);
         this.draw(dot);
         this.emit("change");
       }
 
-      this.drawingImageData = null;
```
Comment (PurpleMyst) [inline js/canvasmanager.ts:120]: Why was this removed?

File: js/drawobjects/circle.ts [hunk 10]
```diff
@@ -25,7 +25,7 @@ export class Circle implements DrawObject {
     return Math.abs(distance(this.center, point) - this.radius);
   }
 
-  getErasingObjects(eraserCenter: Point, eraserRadius: number): DrawObject[] {
+  getObjectsToReplaceWithWhenErasing(eraserCenter: Point, eraserRadius: number): DrawObject[] {
```
Comment (PurpleMyst) [inline js/drawobjects/circle.ts:28]: Think we've gone too far in the other direction, might be a tad bit too verbose. Perhaps `splitWithEraser` and clarify what exactly this means in a JSDoc comment?

**Review by PurpleMyst — CHANGES_REQUESTED** [review]
Very nice set of changes, just the tiniest changes that I've put into individual comments :>

File: js/canvasmanager.ts (hunk 9 above)
Comment (Akuli) [reply js/canvasmanager.ts:120]: I moved it to the beginning of the method in 5c79fb4

**Comment by Akuli** [comment]
note to self: tests are failing, bisect says it's 861ed1f

File: js/drawobjects/circle.ts (hunk 10 above)
Comment (Akuli) [reply js/drawobjects/circle.ts:28]: Why is verbosity bad for this method? The method name is currently written only 4 times in the entire code base.

Comment (PurpleMyst) [reply js/drawobjects/circle.ts:28]: 🤷‍♂️ Guess it doesn't matter all that much.

**Comment by PurpleMyst** [comment]
> 
> 
> note to self: tests are failing, bisect says it's [861ed1f](https://github.com/Akuli/mathpaste/commit/861ed1f261945ab28a4a06bb61110ebfbca55f3c)

Which tests fail?

**Comment by Akuli** [comment]
ci was failing, don't know if there's some way to look at old ci runs in github actions

**Review by PurpleMyst — APPROVED** [review]
LGTM!
